text/0000-disjoins.md

```diff
@@ -0,0 +1,239 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This is a resurrection/expansion of [an older RFC](https://github.com/rust-lang/rust/issues/8277).
+
+Rust supports both kinds of composite algebraic data type: product types
+(structs) and sum types (enums). Rust has anonymous structs (tuples) but no
+anonymous enums, leaving a gap in it's type system.
+
+                    named      anonymous
+                 ------------------------
+                 |           |          |
+        products |  structs  |  tuples  |
+                 |           |          |
+                 ------------------------
+                 |           |          |
+            sums |   enums   |    ??    |
+                 |           |          |
+                 ------------------------
+                 |           |          |
+    exponentials | functions | closures |
+                 |           |          |
+                 ------------------------
+
+This RFC proposes to add anonymous enums to Rust and suggests naming them
+disjoins (as in disjoint unions).
+
+```rust
+let foo: (char | i32 | i32) = (!|!|123);
+match foo {
+  (c|!|!) => println!("char in position zero: {:?}", c),
+  (!|i|!) => println!("i32 in position one: {}", i),
+  (!|!|i) => println!("i32 in position two: {}", i),
+};
+
+let foo: (char|) = ('a'|);
+match foo {
+  (c|)  => println!("char in position zero: {:?}", c),
+};
+
+let foo: ! = panic!("no value");
+match foo {
+};
```

Comment:
> the match is reachable for enum Never { }

No it's not. `enum Never {}` can never exist so a function that returns it is a diverging function.

> But for a diverging function it makes no sense to match on it's result

Yes it does. A diverging function doesn't return a value. ie. it returns the type of no values which can be matched with an empty match statement. The following is currently valid rust code:

```
enum Never {}

fn foo() -> Never {
  panic!("oh no");
}

fn main() {
  let x = foo();
  let y: String = match x {}:
}